### ISSUE TITLE
NoHttpResponseException

### ISSUE THREAD

**Issue by xghrbc1001** [neo4j/neo4j#6178]
we use sdn4 to access neo4j,but sometimes NoHttpResponseException exception occurs. what's the reason? neo4j's version 2.2.5。what config we can use to avoid this exception?otherwise can you give me some advice? thank you.

```
org.neo4j.ogm.session.result.ResultProcessingException: Failed to execute request: {"statements":[{"statement":"match n where id(n) in {0} and n.name is not null and n.name <> '' return id(n) as graphId,n.name as name,n.base as base","parameters":{"0":[2253225345,2289284710,1985905196,87746266]},"resultDataContents":["row"],"includeStats":false}]}
    at org.neo4j.ogm.session.request.DefaultRequest.execute(DefaultRequest.java:105) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    at org.neo4j.ogm.session.request.SessionRequestHandler.execute(SessionRequestHandler.java:99) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    at org.neo4j.ogm.session.request.SessionRequestHandler.execute(SessionRequestHandler.java:64) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    at org.springframework.data.neo4j.repository.query.QueryResultGraphRepositoryQuery$1.apply(QueryResultGraphRepositoryQuery.java:65) ~[spring-data-neo4j-4.0.0.RELEASE.jar:na]
    at org.springframework.data.neo4j.repository.query.QueryResultGraphRepositoryQuery$1.apply(QueryResultGraphRepositoryQuery.java:62) ~[spring-data-neo4j-4.0.0.RELEASE.jar:na]
    at org.neo4j.ogm.session.delegates.TransactionsDelegate.doInTransaction(TransactionsDelegate.java:54) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    at org.neo4j.ogm.session.Neo4jSession.doInTransaction(Neo4jSession.java:422) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    at org.springframework.data.neo4j.repository.query.QueryResultGraphRepositoryQuery.mapToConcreteType(QueryResultGraphRepositoryQuery.java:62) ~[spring-data-neo4j-4.0.0.RELEASE.jar:na]
    at org.springframework.data.neo4j.repository.query.QueryResultGraphRepositoryQuery.execute(QueryResultGraphRepositoryQuery.java:52) ~[spring-data-neo4j-4.0.0.RELEASE.jar:na]
    at org.springframework.data.neo4j.repository.query.GraphRepositoryQuery.execute(GraphRepositoryQuery.java:50) ~[spring-data-neo4j-4.0.0.RELEASE.jar:na]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:454) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:432) ~[spring-data-commons-1.11.0.RELEASE.jar:na]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:98) ~[spring-tx-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:262) ~[spring-tx-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:95) ~[spring-tx-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136) ~[spring-tx-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at com.sun.proxy.$Proxy59.findNameByIds(Unknown Source) ~[na:na]
    at com.xxx.service.neo4j.service.HumanService.humanRelatedCompany(HumanService.java:235) ~[HumanService.class:na]
    at com.xxx.service.neo4j.service.HumanService$$FastClassBySpringCGLIB$$b09a87b7.invoke(<generated>) ~[spring-core-4.0.7.RELEASE.jar:na]
    at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:640) ~[spring-aop-4.0.7.RELEASE.jar:4.0.7.RELEASE]
    at com.xxx.service.neo4j.service.HumanService$$EnhancerBySpringCGLIB$$f240757c.humanRelatedCompany(<generated>) ~[spring-core-4.0.7.RELEASE.jar:na]
    at com.xxx.service.neo4j.provider.impl.Neo4jProviderImpl.humanRelatedCompany(Neo4jProviderImpl.java:342) ~[Neo4jProviderImpl.class:na]
    at com.alibaba.dubbo.common.bytecode.Wrapper21.invokeMethod(Wrapper21.java) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.proxy.javassist.JavassistProxyFactory$1.doInvoke(JavassistProxyFactory.java:46) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.proxy.AbstractProxyInvoker.invoke(AbstractProxyInvoker.java:72) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.InvokerWrapper.invoke(InvokerWrapper.java:53) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.ExceptionFilter.invoke(ExceptionFilter.java:64) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.TimeoutFilter.invoke(TimeoutFilter.java:42) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.monitor.support.MonitorFilter.invoke(MonitorFilter.java:65) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.dubbo.filter.TraceFilter.invoke(TraceFilter.java:78) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.ContextFilter.invoke(ContextFilter.java:70) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.GenericFilter.invoke(GenericFilter.java:132) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.ClassLoaderFilter.invoke(ClassLoaderFilter.java:38) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.filter.EchoFilter.invoke(EchoFilter.java:38) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.ProtocolFilterWrapper$1.invoke(ProtocolFilterWrapper.java:91) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.rpc.protocol.dubbo.DubboProtocol$1.reply(DubboProtocol.java:113) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.remoting.exchange.support.header.HeaderExchangeHandler.handleRequest(HeaderExchangeHandler.java:84) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.remoting.exchange.support.header.HeaderExchangeHandler.received(HeaderExchangeHandler.java:170) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.remoting.transport.DecodeHandler.received(DecodeHandler.java:52) [dubbo-2.8.4.jar:2.8.4]
    at com.alibaba.dubbo.remoting.transport.dispatcher.ChannelEventRunnable.run(ChannelEventRunnable.java:82) [dubbo-2.8.4.jar:2.8.4]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_71]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_71]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_71]
Caused by: org.apache.http.NoHttpResponseException: 10.171.23.230:7475 failed to respond
    at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:143) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:261) ~[httpcore-4.4.1.jar:4.4.1]
    at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:165) ~[httpcore-4.4.1.jar:4.4.1]
    at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:167) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:272) ~[httpcore-4.4.1.jar:4.4.1]
    at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:124) ~[httpcore-4.4.1.jar:4.4.1]
    at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:271) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82) ~[httpclient-4.4.1.jar:4.4.1]
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107) ~[httpclient-4.4.1.jar:4.4.1]
    at org.neo4j.ogm.session.request.DefaultRequest.execute(DefaultRequest.java:80) ~[neo4j-ogm-1.1.4-20151201.162656-1.jar:na]
    ... 56 common frames omitted
```

**Comment by tjwde** [comment]
I have the same issue using spring-data-neo4j (4.0.0.RELEASE) together with neo4j-ogm (1.1.4) and Neo4j 2.3.1 Server (Windows 7 environment / test / developer workstation).

**Comment by daviduck123** [comment]
Yesterday, i don't have a problem, just start the Neo4j Server and run the application and its work.
Today, i did the same, but get this error.

**Comment by vince-bickers** [comment]
We're currently investigating what the problem could be. 

Please also see these questions on SO:

http://stackoverflow.com/questions/34606163/why-nohttpresponseexception-occues-while-using-sdn4/34608737#34608737

http://stackoverflow.com/questions/24752278/org-apache-http-nohttpresponseexception-the-target-server-failed-to-respond?lq=1

There are two observations I can make at the moment.
1. The problem is intermittent, so it is very hard to diagnose.
2. The problem has not been reported until quite recently.

The second observation may or may not be relevant. It could simply be that as more people use SDN 4/OGM the probability of this problem occurring is higher, and we’re seeing that. But I don’t want to rule out the possibility that we’ve done something somewhere (e.g. in handling timed-out transactions) that has introduced a defect elsewhere.

**Comment by vince-bickers** [comment]
All,

The latest snapshot version of the OGM, 1.1.5-SNAPSHOT has a request retry strategy enabled in the event that a NoHttpResponseException is thrown: https://github.com/neo4j/neo4j-ogm 

The retry strategy is configured to attempt the request 3 more times after the initial failure, with a delay of 2 seconds between each attempt. 

The rationale for this strategy is based on the Apache HttpClient documentation which states the following:

> In some circumstances, usually when under heavy load, the web server may be able to receive requests but unable to process them. A lack of sufficient resources like worker threads is a good example. This may cause the server to drop the connection to the client without giving any response. HttpClient throws NoHttpResponseException when it encounters such a condition. In most cases it is safe to retry a method that failed with NoHttpResponseException.
> <cite>http://hc.apache.org/httpclient-3.x/exception-handling.html</cite>

The 1.1.5-SNAPSHOT version of the OGM is compatible with SDN 4.0. 

**OGM users**
If you're using the OGM "standalone", you'll need to add this dependency to your pom:

```
   <repository>
        <id>neo4j-snapshots</id>
        <url>http://m2.neo4j.org/content/repositories/snapshots</url>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
    </repository>
```

**SDN users**
The dependency should already be included in the latest Spring Data 4.0.0-SNAPSHOT build. You'll need the following repository in your pom to use the snapshot build of SDN 4.0

```
    <repository>
        <id>spring-libs-snapshot</id>
        <url>http://repo.spring.io/libs-snapshot</url>
    </repository>
```

**Comment by technige** [comment]
Old issue: closing. Please open a new issue if required.
